### PR TITLE
Add exact name ban option

### PR DESCRIPTION
**Pull request by 12pm**
To entirely disallow a name

### CODE REVIEW

**Comment by Learath2** [comment]
I think a distance of 0 implies an exact match.

**Comment by 12pm** [comment]
Distance of 0 means to not check for similar names or no? Exact bans everything that contains the banned name, so more strict

File: src/engine/server/name_ban.h [hunk 1]
```diff
@@ -14,18 +14,20 @@ class CNameBan
...
 public:
 	CNameBan() {}
-	CNameBan(const char *pName, int Distance, const char *pReason = "") :
+	CNameBan(const char *pName, int Distance, int Exact, const char *pReason = "") :
 		m_Distance(Distance)
 	{
 		str_copy(m_aName, pName, sizeof(m_aName));
 		m_SkeletonLength = str_utf8_to_skeleton(m_aName, m_aSkeleton, sizeof(m_aSkeleton) / sizeof(m_aSkeleton[0]));
 		str_copy(m_aReason, pReason, sizeof(m_aReason));
+		m_Exact = Exact;
```
Comment (heinrich5991) [inline src/engine/server/name_ban.h]: Can be initialized like `m_Distance`.

**Comment by 12pm** [comment]
> General comment on effectivity: If I think that I want to "circumvent" a name ban on `jao` and I can't choose `*jao*` anymore, I guess I'll just take `j.a.o`.

It should help against people adding a prefix/suffix to their banned name, which is most convenient to stay recognizable. Happened in the past.
 
> I think "exact" is not describing what the code does, we should find a better name.

Any suggestion? "strict"?

**Comment by heinrich5991** [comment]
> Any suggestion? "strict"?

Perhaps "substring"? "strict" and "exact" only communicate that it's stronger than before, but that is not really meaningful if you haven't been involved in the development of the feature.

**Comment by 12pm** [comment]
Hm that sounds like you have to pass a string

**Comment by def-** [comment]
Then is_substring

**Comment by Learath2** [comment]
Shouldn't maybe all name bans behave this way?

**Comment by heinrich5991** [comment]
> Shouldn't maybe all name bans behave this way?

I thought the same, I dislike having more options that we need to account for.

This however, would make it hard to ban someone called `A` or similar.

**Comment by 12pm** [comment]
We banned someone with a 2 character name, that would have been problematic if this applied to every ban

**Comment by def-** [comment]
I'd say let's get this in, see if it's used much and then we can think about it further. Ok?

**Comment by heinrich5991** [comment]
I would like this to have a name that says something (it doesn't have to be super specific) first.

The current name is akin to naming a function `ProcessData`.

**Comment by 12pm** [comment]
What do you think about is_substring @heinrich5991

**Comment by heinrich5991** [comment]
Works.

**Comment by def-** [comment]
@12pm Want to do the rename or should to `is_substring` or should I?

**Comment by heinrich5991** [comment]
Thanks!

bors r+

**Pull request merged**